### PR TITLE
feat: add `redef` keyword

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to https://github.com/flix/flix/issues/8268

### CODE REVIEW

**Comment by jaschdoc** [comment]
Should I just add another function to the parser called `redefinitionDecl` instead of extending the `definitionDecl` function and doing if-checks? Seems like a code smell

**Comment by magnus-madsen** [comment]
> Should I just add another function to the parser called `redefinitionDecl` instead of extending the `definitionDecl` function and doing if-checks? Seems like a code smell

No, I think we should try do so a simple desugaring so that most the compiler stays the same.

**Comment by magnus-madsen** [comment]
I suppose your question is: where should you do the desugaring? which is a good question.

**Comment by jaschdoc** [comment]
Hmm we do have a separate desugaring phase so maybe there?

**Comment by magnus-madsen** [comment]
> Hmm we do have a separate desugaring phase so maybe there?

I think we could, yes.

File: main/src/ca/uwaterloo/flix/language/phase/Desugar.scala [hunk 1]
```diff
@@ -120,6 +123,22 @@ object Desugar {
...
+    case WeededAst.Declaration.Redef(doc, ann, mod, ident, tparams0, fparams0, exp0, tpe0, eff0, tconstrs0, econstrs0, loc) =>
+      flix.subtask(ident.name, sample = true)
+      val tparams = tparams0.map(visitTypeParam)
+      val fparams = visitFormalParams(fparams0)
+      val exp = visitExp(exp0)
+      val tpe = visitType(tpe0)
+      val eff = eff0.map(visitType)
+      val tconstrs = tconstrs0.map(visitTraitConstraint)
+      val econstrs = econstrs0.map(visitEqualityConstraint)
+      DesugaredAst.Declaration.Def(doc, ann, mod, ident, tparams, fparams, exp, tpe, eff, tconstrs, econstrs, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Desugar.scala:139]: I think here we should add the override modifier, yes?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Desugar.scala:139]: Yeah oops forgot that

**Pull request merged**